apollo-compiler/src/main/kotlin/com/apollographql/apollo3/compiler/unified/IrBuilder.kt

```diff
@@ -0,0 +1,318 @@
+
+    val dataFieldResult = createDataField(selectionSet, typeDefinition.name)
+
+    val variableDefinitions = inferVariables(schema, allGQLFragmentDefinitions)
+
+    return IrNamedFragment(
+        name = name,
+        description = description,
+        filePath = sourceLocation.filePath!!,
+        typeCondition = typeDefinition.name,
+        variables = variableDefinitions.map { it.toIr() },
+        dataField = dataFieldResult.field,
+    )
+  }
+
+  class DataFieldResult(
+      val field: IrField,
+      val usedNamedFragments: Set<String>,
+  )
+
+  private fun createDataField(selectionSet: GQLSelectionSet, typeCondition: String): DataFieldResult {
+    val usedNamedFragments = mutableSetOf<String>()
+
+    val irFieldSets = IrFieldSetBuilder(
+        schema,
+        allGQLFragmentDefinitions,
+        selectionSet.selections,
+        typeCondition,
+        { usedNamedFragments.add(it) },
+        { it.toIr() }
+    ).build()
+
+    val field = IrField(
+        name = "data",
+        alias = null,
+        deprecationReason = null,
+        arguments = emptyList(),
+        type = IrObjectType(typeCondition),
+        condition = BooleanExpression.True,
+        description = "Synthetic data field",
+        fieldSets = irFieldSets,
+    )
+
+    return DataFieldResult(field, usedNamedFragments)
+  }
+
+  private fun InputValueScope.VariableReference.toIr(): IrVariable {
+    return IrVariable(
+        name = this.variable.name,
+        defaultValue = null,
+        type = expectedType.toIr(),
+    )
+  }
+
+  private fun GQLVariableDefinition.toIr(): IrVariable {
+    val coercedDefaultValue = defaultValue?.coerce(type, schema)?.orThrow()
+
+    return IrVariable(
+        name = name,
+        defaultValue = coercedDefaultValue?.toIr(),
+        type = type.toIr(),
+    )
+  }
+
+  /**
+   * Maps to [IrType] and also keep tracks of what types are actually used so we only generate those
+   */
+  private fun GQLType.toIr(): IrType {
+    return when (this) {
+      is GQLNonNullType -> IrNonNullType(ofType = type.toIr())
+      is GQLListType -> IrListType(ofType = type.toIr())
+      is GQLNamedType -> when (schema.typeDefinition(name)) {
+        is GQLScalarTypeDefinition -> {
+          when (name) {
+            "String" -> IrStringType
+            "Boolean" -> IrBooleanType
+            "Int" -> IrIntType
+            "Float" -> IrFloatType
+            "ID" -> IrIdType
+            else -> {
+              usedCustomScalars.add(name)
+              IrCustomScalarType(name)
+            }
+          }
+        }
+        is GQLEnumTypeDefinition -> {
+          usedEnums.add(name)
+          IrEnumType(name)
+        }
+        is GQLObjectTypeDefinition -> IrObjectType(name)
+        is GQLInterfaceTypeDefinition -> IrInterfaceType(name)
+        is GQLUnionTypeDefinition -> IrUnionType(name)
+        is GQLInputObjectTypeDefinition -> {
+          usedInputObjects.add(name)
```

Comment:
This is where we record the used types. This way we have a central place to record all usages.